Sources/Algorithms/Trim.swift

```diff
@@ -0,0 +1,48 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension BidirectionalCollection {
+
+  /// Returns a `SubSequence` formed by discarding all elements at the start and end of the Collection
+  /// which satisfy the given predicate.
+  ///
+  /// Trimming is a common operation in computing, often referring to removing leading and trailing whitespace
+  /// from a string[1].
+  ///
+  ///  ```
+  ///  let myString = "  hello, world  "
+  ///  print(myString.trimmed(where: \.isWhitespace)) // "hello, world"
+  ///  ```
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Trimming_(computer_programming)
+  ///
+  /// - parameters:
+  ///    - predicate:  A closure which determines if the element should be omitted from the resulting slice.
+  ///
+  /// - complexity: `O(n)`, where `n` is the length of the Collection.
+  ///
+  @inlinable
+  public func trimmed(where predicate: (Element) throws -> Bool) rethrows -> SubSequence {
+
+    // Consume elements from the front.
+    let sliceStart = try firstIndex(where: { try predicate($0) == false }) ?? endIndex
+    // sliceEnd is the index _after_ the last index to match the predicate.
+    var sliceEnd = endIndex
+    while sliceStart != sliceEnd {
+      let idxBeforeSliceEnd = index(before: sliceEnd)
+      guard try predicate(self[idxBeforeSliceEnd]) else {
+        return self[sliceStart..<sliceEnd]
```

Comment:
Q: can these bounds be safely unchecked too?

Reply:
I'm not 100% sure. Since `sliceEnd` is formed by walking backwards from `endIndex` and has not yet tested equal to `sliceStart`, the only way this bounds check will fail is if either your collection or the index's `Comparable` conformance is broken.

For the empty result, we already checked that `sliceStart == sliceEnd` as part of the `while` loop condition, so it is safe to omit the bounds check. A broken conformance would never get to that part.